### PR TITLE
add `c_str!` macro to create `&'static CStr`

### PR DESCRIPTION
**Pull request by davidhewitt**
With Rust 1.77 finally stabilising `c""` c-string literals, I've wanted for a little while to be able to make better use of `&'static CStr` inside PyO3. 

This PR proposes a pathway to doing that, by adding a `c_str!("foo")` macro to create a static c-string safely. With MSRV 1.63 we're able to support this on all versions.

This has the nice effect of moving work from runtime to compile time, because at the moment we have a lot of `&str` which need to be converted to `&CStr` using runtime checks. In the future when MSRV passes 1.77 we have a trivial cleanup to swap everything to `c""` literals.

The only user-facing changes:
- Add `pyo3_ffi::c_str` macro (also accessible as `pyo3::ffi::c_str`).
- The `PyCFunction` constructors now take `&'static CStr` where they previously took `&'static str`.

### CODE REVIEW

File: src/impl_/pymethods.rs [hunk 1]
```diff
@@ -284,22 +272,6 @@ where
     retval
 }
 
-pub(crate) struct PyMethodDefDestructor {
```
Comment (mejrs) [inline src/impl_/pymethods.rs:287]: I'm happy to get rid of this :)

**Review by mejrs — COMMENTED** [review]
I'm a big fan of this :)

File: examples/sequential/src/id.rs [hunk 2]
```diff
@@ -27,10 +28,7 @@ unsafe extern "C" fn id_new(
     kwds: *mut PyObject,
 ) -> *mut PyObject {
     if PyTuple_Size(args) != 0 || !kwds.is_null() {
-        PyErr_SetString(
-            PyExc_TypeError,
-            "Id() takes no arguments\0".as_ptr().cast::<c_char>(),
-        );
+        PyErr_SetString(PyExc_TypeError, c_str!("Id() takes no arguments").as_ptr());
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply examples/sequential/src/id.rs:31]: Good idea!

While having the example use `c""` is tempting, I think it's simpler for CI and also more consistent if we use the macro for now. I look forward to 1-2 years from now when we have MSRV able to support this!

File: src/impl_/pymethods.rs (hunk 1 above)
Comment (davidhewitt) [reply src/impl_/pymethods.rs:287]: Me too! It felt so awkward before to have to do the dance with `&'static str` in a bunch of places. In some we created these destructors and in others we made an unsafe requirement of no interior nulls. This end state is much more satisfying.

File: pyo3-macros-backend/src/pymethod.rs [hunk 3]
```diff
@@ -870,7 +870,8 @@ pub enum PropertyType<'a> {
 }
 
 impl PropertyType<'_> {
-    fn null_terminated_python_name(&self) -> Result<syn::LitStr> {
+    fn null_terminated_python_name(&self, ctx: &Ctx) -> Result<TokenStream> {
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply pyo3-macros-backend/src/pymethod.rs:873]: I'm happy to go with that; I might also see how bad it is to use conditional code to emit `c""` directly from macros on new enough versions. I think it'd be much more efficient codegen and makes it easier for us to clean up later.

... but maybe in the interest of not blocking the 0.22 release too much, would you be ok if I punt doing this into a follow-up internal refactoring?

[the remark this answers is not in the document]
Comment (mejrs) [reply pyo3-macros-backend/src/pymethod.rs:873]: Sure 👍

**Comment by Icxolu** [comment]
Big 👍 for this from me as well! This is a lot nicer and safer than having to remember to add the `\0` and it also gets rid of a lot of casting which is nice to see.

**Comment by davidhewitt** [comment]
Given the two broadly positive sentiments, I think I'm going to merge this once I get CI passing. It'll make #4254 easier to finish implementing, which I want to complete ASAP with a view to then shipping 0.22.

If there are changes which are wanted (like the `PythonDoc` changes proposed above) let's put them into follow-up PRs or issues to land after we've shipped and unblocked everyone downstream from 3.13 testing.

**Pull request merged**